### PR TITLE
[Version 3.6] Add GenerateCoverageViz pipeline step to dag.

### PR DESCRIPTION
**Pull request by MarkAZhang**
Also slight change to BlastContigs pipeline step to generate blast.top.m8 file.

Testing steps:
* Tested the benchmark sample with this dag branch on staging, and it succeeded as expected. https://staging.idseq.net/samples/12576
* Tested a minimal sample with no assembled contigs with this dag branch on staging, and it also succeeded (this tests the case where some files are empty) https://staging.idseq.net/samples/12575

The coverage viz displays correctly using the files generated by this pipeline step.

Step takes about 1 minute to run for the benchmark sample, based on cloudwatch logs.

### CODE REVIEW

File: examples/assemble_blast.json [hunk 1]
```diff
@@ -28,13 +28,15 @@
     "rapsearch2_accessions_out": ["nr.refseq.fasta"],
     "refined_gsnap_out": [
       "gsnap.blast.m8",
+      "gsnap.blast.top.m8",
```
Comment (yunfangjuan) [inline examples/assemble_blast.json]: move this to the end of the json?

File: idseq_dag/steps/blast_contigs.py [hunk 2]
```diff
@@ -30,7 +30,7 @@ def run(self):
         assembled_contig, _assembled_scaffold, bowtie_sam, _contig_stats = self.input_files_local[1]
         reference_fasta = self.input_files_local[2][0]
 
-        (blast_m8, refined_m8, refined_hit_summary, refined_counts, contig_summary_json) = self.output_files_local()
+        (blast_m8, top_entry_m8, refined_m8, refined_hit_summary, refined_counts, contig_summary_json) = self.output_files_local()
```
Comment (yunfangjuan) [inline idseq_dag/steps/blast_contigs.py]: top entry to the end of the output?

File: idseq_dag/steps/download_accessions.py [hunk 3]
```diff
@@ -122,5 +149,21 @@ def fetch_sequence_for_thread(error_flags, accession, accession_out_file,
...
+
+    def generate_size_file(self, accession_summary_tab, accession_dict, loc_db, accession_summary):
+        loc_dict = open_file_db_by_extension(loc_db, IdSeqDictValue.VALUE_TYPE_ARRAY)
 
+        with open(accession_summary_tab, 'w') as ast:
+            for accession_id, _ in accession_dict.items():
+                entry = loc_dict.get(accession_id)
+                if entry and len(entry) == 3:
+                    # The seq_len includes newline characters in the fasta file, so we must correct for this.
+                    seq_len = math.floor(FASTA_SEQ_LINE_LENGTH / (FASTA_SEQ_LINE_LENGTH + 1) * int(entry[2]))
```
Comment (yunfangjuan) [inline idseq_dag/steps/download_accessions.py]: hm it's still not clear what this is.  also I think it's a bit dangerous to assume FASTA_SEQ_LINE_LENGTH is 80. Can you just make this a rough length?

File: idseq_dag/steps/download_accessions.py [hunk 4]
```diff
@@ -79,14 +94,26 @@ def download_ref_sequences_from_s3(self, accession_dict, output_reference_fasta,
             t.join()
         if error_flags:
             raise RuntimeError("Error in getting sequences by accession list.")
+
+        for accession, taxinfo in accession_dict.items():
+            accession_out_file = os.path.join(accession_dir, accession)
+            cmd = f"cat {accession_out_file} |head -1"
+            seq_name = subprocess.check_output(
+                    cmd, executable='/bin/bash', shell=True).decode("utf-8").split(" ", 1)[1].replace("\n", "")
+            accession_summary[accession]["name"] = seq_name
```
Comment (yunfangjuan) [inline idseq_dag/steps/download_accessions.py]: this might actually not be thread-safe though. Also, it's probably easier to get the sequence length here?

File: idseq_dag/steps/download_accessions.py [hunk 5]
```diff
@@ -122,5 +149,21 @@ def fetch_sequence_for_thread(error_flags, accession, accession_out_file,
         finally:
             semaphore.release()
 
+    def generate_dummy_fasta_ref(self, output_reference_fasta):
```
Comment (yunfangjuan) [inline idseq_dag/steps/download_accessions.py]: what's this function for?

File: idseq_dag/steps/download_accessions.py (hunk 3 above)
Comment (MarkAZhang) [reply idseq_dag/steps/download_accessions.py]: This calculates the number of nucleotide bases, and removes the newline characters from the count.

Good point about assuming 80. I will see if I can calculate dynamically in some way.

File: idseq_dag/steps/download_accessions.py (hunk 5 above)
Comment (MarkAZhang) [reply idseq_dag/steps/download_accessions.py]: WIP placeholder function.

File: idseq_dag/steps/download_accessions.py (hunk 4 above)
Comment (MarkAZhang) [reply idseq_dag/steps/download_accessions.py]: What would make something not thread-safe? Not sure what I should be thinking about.

File: idseq_dag/steps/download_accessions.py (hunk 3 above)
Comment (cdebourcy) [reply idseq_dag/steps/download_accessions.py]: If I understand correctly, the issue is that we cannot use the size of the byterange as the sequence length, because byterange includes new-line characters. I can see a few different options, some of which are much better than others:

(a) Calculate all the sequence lengths from the downloaded sequence files. This would be super wasteful to do at pipeline run time.
(b) Determine the actual sequence length while generating nt_loc_db (this is part of the index update operation) and store it in the dict once and for all so it can just be looked up. Probably the best option, but requires the development/testing overhead of building a new nt_loc_db.
(c) Keep the current setup with assumed line length of 80, but do a quick spot check in the code to confirm sequence files actually satisfy this condition. This code can probably live in the step that generates a new nt_loc_db when we update the index. Include an assertion with a very explicit / verbose message for the developers if the condition fails, so that they know how to address the problem (for example, "you need to change the value of FASTA_SEQ_LINE_LENGTH in download_accessions.py"). For what it's worth, I do think the line length of 80 will be a good assumption for a long time to come. Wikipedia article says it's something of a standard: https://en.wikipedia.org/wiki/FASTA_format

Anything I missed?

**Comment by yunfangjuan** [comment]
I think b) is good and is actually pretty easy to do. You change the step,
calculate length and output it to a new sqlite3 file. It can also serve
other purposes if people are trying to curate NT/NR.

File: examples/assemble_blast.json (hunk 1 above)
Comment (cdebourcy) [reply examples/assemble_blast.json]: Agreed, it might be less prone to human error if new outputs are added to the end of the list rather than in the middle. Or just confirm that all the steps that refer to `refined_gsnap_out` still refer to the correct indices of these lists. For example, the postprocess dag in idseq-web (https://github.com/chanzuckerberg/idseq-web/blob/7da56caff3a686bcf2559c51ece28a0f258779c5/app/lib/dags/postprocess.json.erb#L146) indicates that the step combine_taxon_counts (https://github.com/chanzuckerberg/idseq-dag/blob/master/idseq_dag/steps/combine_taxon_counts.py#L12) is pulling the item at index 3 of the `refined_gsnap_out` list as an input. But your change shifts the indices of this list, so it would break

Comment (cdebourcy) [reply examples/assemble_blast.json]: [Note to self for future: It might be a good idea to change the targets into dicts so we can use named keys and stop referring to target items by their index in the code for the pipeline steps. Might speed up development]

File: idseq_dag/steps/download_accessions.py (hunk 4 above)
Comment (cdebourcy) [reply idseq_dag/steps/download_accessions.py]: @yunfangjuan do you want to elaborate? This happens after the threads are joined, no?

Comment (yunfangjuan) [reply idseq_dag/steps/download_accessions.py]: Writing to the same dict simultaneously

Comment (cdebourcy) [reply idseq_dag/steps/download_accessions.py]: To me it looks as if all the download threads are joined in line 94 (https://github.com/chanzuckerberg/idseq-dag/blob/ffef5bf9a6aa7b0284775d9f11a7818e28dc4577/idseq_dag/steps/download_accessions.py#L94) and writing to the dict is single-threaded.

The two places in the code that write to `accession_summary`, namely the methods `download_ref_sequences_from_s3` and `get_sequence_by_accession_from_file`, are called in a mutually exclusive way in the code.

@yunfangjuan Anything I'm missing? Please advise.

Comment (yunfangjuan) [reply idseq_dag/steps/download_accessions.py]: OK. This case it's actually done after the join so there's only one thread doing it, so it's fine. But I think a better way to deal with this will be to generate another dictionary from the generate_loc_db step to pull out the seq_name and sequence length once for each index update so you won't add time to our current pipeline.

File: tests/generate_coverage_viz.py [hunk 6]
```diff
@@ -0,0 +1,344 @@
+import unittest
+
+from idseq_dag.steps.generate_coverage_viz import PipelineStepGenerateCoverageViz
+
+class GenerateCoverageViz(unittest.TestCase):
```
Comment (cdebourcy) [inline tests/generate_coverage_viz.py]: Love these tests!

File: idseq_dag/steps/generate_coverage_viz.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+                    "percent_id": float(parts[2]),
+                    "alignment_length": int(parts[3]),
+                    "num_mismatches": int(parts[4]),
+                    "num_gaps": int(parts[5])
+                }
+
+        return reads
+
+
+    def augment_contigs_map_with_coverage(self, contig_coverage_json, contigs_map):
```
Comment (yunfangjuan) [inline idseq_dag/steps/generate_coverage_viz.py]: Do you need self? Can make this s staticmethod?

**Comment by yunfangjuan** [comment]
In general, this is rad!!

File: idseq_dag/steps/generate_coverage_viz.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+        # Create a coverage viz JSON file for each accession
+        coverage_viz_dir = os.path.join(self.output_dir_local, "coverage_viz")
+        command.execute(f"mkdir -p {coverage_viz_dir}")
+        for accession_id in coverage_viz_obj:
+            upload_file = os.path.join(coverage_viz_dir, f"{accession_id}_coverage_viz.json")
+
+            with open(upload_file, 'w') as uf:
+                json.dump(coverage_viz_obj[accession_id], uf)
+
+            self.additional_files_to_upload.append(upload_file)
```
Comment (cdebourcy) [inline idseq_dag/steps/generate_coverage_viz.py]: This needs to be changed to use `additional_folders_to_upload` instead of `additional_files_to_upload`. The former is a mechanism I just introduced in my last PR. It uploads everything in the folder to S3 in a single AWS CLI request. The latter uploads every file using an individual request. For thousands and thousands of files it's going to take hours.

Comment (cdebourcy) [reply idseq_dag/steps/generate_coverage_viz.py]: Please review the optimizations I introduced to cut the experimental stage from 3 days to 3 hours here: https://github.com/chanzuckerberg/idseq-dag/pull/139

File: idseq_dag/steps/generate_coverage_viz.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    def augment_contigs_map_with_coverage(self, contig_coverage_json, contigs_map):
+        with open(contig_coverage_json, 'r') as ccj:
+            contig_coverage = json.load(ccj)
+
+            for contig_name in contig_coverage:
+                if contig_name in contigs_map:
+                    contigs_map[contig_name]["coverage"] = contig_coverage[contig_name]["coverage"]
+
+
+    def augment_contigs_map_with_byteranges(self, contigs_fasta, contigs_map):
```
Comment (cdebourcy) [inline idseq_dag/steps/generate_coverage_viz.py]: Yay!

File: idseq_dag/steps/generate_coverage_viz.py [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+        if os.path.getsize(deduped_m8) < MIN_M8_FILE_SIZE:
+            return contigs
+
+        # iterate_m8 automatically removes invalid hits.
+        for _read_id, _accession_id, _percent_id, _alignment_length, _e_value, _bitscore, line in m8.iterate_m8(deduped_m8):
+            parts = line.split("\t")
+
+            if parts[0] in unassigned_reads_set:
+                reads[parts[0]] = {
+                    # Extract the length from the contig name.
```
Comment (cdebourcy) [inline idseq_dag/steps/generate_coverage_viz.py]: Why the code duplication? Can make a helper to parse the general m8 into this structure

File: idseq_dag/steps/generate_coverage_viz.py [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+
+        # iterate_m8 automatically removes invalid hits.
+        for _contig_id, _accession_id, _percent_id, _alignment_length, _e_value, _bitscore, line in m8.iterate_m8(blast_top_m8):
+            parts = line.split("\t")
+            name_parts = parts[0].split("_")
+
+            if parts[0] in valid_contigs_with_read_counts:
+                contigs[parts[0]] = {
+                    "total_length": int(name_parts[3]),
+                    "accession": parts[1],
```
Comment (cdebourcy) [inline idseq_dag/steps/generate_coverage_viz.py]: duplicated with line 199

**Comment by cdebourcy** [comment]
Also, could you please test a few large, real-life samples? In order to assess performance and results for different distributions of reads across contigs and accessions.

The benchmark has a low number of reads and was generated from a small number of species covered really really well which makes it easy to assemble them into small numbers of contigs. Real samples will have a much much longer tail of singletons

**Comment by MarkAZhang** [comment]
Regarding utils vs static methods, it's pretty arbitrary.

The static methods are mostly reading files, and the util methods are calculations and JSON object compilation.

One other distinction is the util methods are all tested with unit tests, and the static methods aren't since I wasn't sure how to mock files.

**Comment by MarkAZhang** [comment]
Talked to @yunfangjuan today about this PR.

She had two concerns:
* I wasn't taking paired reads into account, and she remembered some tricky logic in alignment-viz and `ReadViz.jsx` involving paired reads. I dug into this, and it turns out paired reads can affect the indices for the read, but not for the reference sequence. Since I only need the indices for the reference sequence to figure out where to add coverage to account for the read, I don't need to take paired reads into account for coverage viz. (in alignment viz, we also need to extract the exact aligned sequence from the read, in the correct orientation, which is why it's important)
* She was concerned about my putting a lot of functions into the `utils` folder that were specific to the CoverageViz step. She was also concerned that my code was a bit hard to read because I didn't group my functions together well enough. I'll do a pass to package/modularize my functions into neater modules with clear inputs/outputs.

**Comment by MarkAZhang** [comment]
Latest commit:
Improve code organization for coverage viz pipeline step.
    
 Removed util/coverage and moved functions into PipelineStepGenerateCoverageViz,
 since those functions are specific to that one pipeline step.
    
 Reorganized functions in PipelineStepGenerateCoverageViz so that the "main"
 functions are at the top of the file.
  
 Consolidated tests into one file.

Verified that:
* Tests still pass.
* The pipeline step still outputs the exact same JSON as before.

File: idseq_dag/steps/generate_coverage_viz.py [hunk 12]
```diff
@@ -0,0 +1,823 @@
...
+        with open(contig_stats_json, 'r') as csj:
+            contig_read_counts = json.load(csj)
+
+            for contig in contig_read_counts:
+                if contig != "*" and contig_read_counts[contig] >= min_contig_size:
+                    valid_contigs_with_read_counts[contig] = contig_read_counts[contig]
+
+        return valid_contigs_with_read_counts
+
+    # Generate a dict that maps accessions to the reads and contigs that were assigned to them.
```
Comment (yunfangjuan) [inline idseq_dag/steps/generate_coverage_viz.py]: I believe for the python convention you will put the documentation inside the function with ''' ''' around it.

File: idseq_dag/steps/generate_coverage_viz.py [hunk 13]
```diff
@@ -0,0 +1,823 @@
...
+            "avg_prop_mismatch": prop_total_mismatch / (len(accession_data["contigs"]) + len(accession_data["reads"]))
+        }
+
+
+# Private utility methods used by the pipeline step.
+
+# Decrement the lower bound of a possibly inverted interval.
+# i.e. turn (1, 5) into (0, 5) and (5, 1) into (5, 0).
+# Intervals in m8 alignment files can be inverted.
+def _decrement_lower_bound(interval):
```
Comment (yunfangjuan) [inline idseq_dag/steps/generate_coverage_viz.py]: I like it.

File: idseq_dag/steps/generate_coverage_viz.py (hunk 12 above)
Comment (MarkAZhang) [reply idseq_dag/steps/generate_coverage_viz.py]: Done.

File: idseq_dag/steps/generate_coverage_viz.py [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+import idseq_dag.util.command as command
+from idseq_dag.engine.pipeline_step import PipelineStep
+# coverage_utils contains the bulk of the business logic for this pipeline step
+import idseq_dag.util.coverage as coverage_utils
+from idseq_dag.util.dict import IdSeqDictValue, open_file_db_by_extension
+
+MIN_M8_FILE_SIZE = 25 # M8 file should have at least a single line.
+
+# These can be overridden with additional_attributes
+MAX_NUM_BINS_COVERAGE = 500
```
[the remark this answers is not in the document]
Comment (MarkAZhang) [reply idseq_dag/steps/generate_coverage_viz.py]: Added comment.
